src/api/c/anisotropic_diffusion.cpp

```diff
@@ -0,0 +1,97 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/dim4.hpp>
+#include <af/image.h>
+#include <handle.hpp>
+#include <common/err_common.hpp>
+#include <backend.hpp>
+#include <arith.hpp>
+#include <cast.hpp>
+#include <copy.hpp>
+#include <gradient.hpp>
+#include <reduce.hpp>
+#include <anisotropic_diffusion.hpp>
+
+#include <type_traits>
+
+using af::dim4;
+using namespace detail;
+
+template<typename T>
+af_array diffusion(const Array<T> in, const float dt, const float K,
+                   const unsigned iterations, const af_flux_function fftype,
+                   const af::diffusionEq eq)
+{
+    auto out    = copyArray(in);
+    auto dims   = out.dims();
+    auto g0     = createEmptyArray<T>(dims);
+    auto g1     = createEmptyArray<T>(dims);
+    float cnst  = -2.0*K*K/dims.elements();
+
+    for (unsigned i=0; i<iterations; ++i)
+    {
+        gradient<T>(g0, g1, out);
+
+        auto g0Sqr = arithOp<T, af_mul_t>(g0, g0, dims);
+        auto g1Sqr = arithOp<T, af_mul_t>(g1, g1, dims);
+        auto sumd  = arithOp<T, af_add_t>(g0Sqr, g1Sqr, dims);
+
+        T avg = reduce_all<af_add_t, T, T>(sumd, true, 0);
+
+        anisotropicDiffusion(out, dt, 1.0/(cnst*avg), fftype, eq);
+    }
+
+    return getHandle(out);
+}
+
+af_err af_anisotropic_diffusion(af_array* out, const af_array in, const float dt,
+                                const float K, const unsigned iterations,
+                                const af_flux_function fftype,
+                                const af_diffusion_eq eq)
+{
+    try {
+        const ArrayInfo& info = getInfo(in);
+
+        const af::dim4& inputDimensions = info.dims();
+        const af_dtype  inputType       = info.getType();
+        const unsigned  inputNumDims    = inputDimensions.ndims();
+
+        DIM_ASSERT(1, (inputNumDims>=2));
+
+        ARG_ASSERT(3, (K>0 || K<0));
+        ARG_ASSERT(4, (iterations>0));
+
+        float DT = dt;
+        float maxDt = 1.0f/std::pow(2.0f, static_cast<float>(2)+1);
+#ifndef NDEBUG
+        if (dt > maxDt) {
+            std::cerr<< "WARNING: maximum time step for 2D Image is "<< maxDt << std::endl;
```

Comment:
is this supposed to be an assert?

Reply:
No, we don't need to prevent the execution because of timestep beyond the stable range. The resulting image might not be accepatable in that particular domain. However, if some beyond range timestep works for an user, we should let them run it. Should we remove it ?

Reply:
I don't want ArrayFire to print anything unless it is something the user wants to print like af::info